apps/core/secureaction.cpp

```diff
@@ -0,0 +1,254 @@
+/****************************************************************************
+**
+** Copyright (c) 2018 - 2021 Jolla Ltd.
+** Copyright (c) 2021 Open Mobile Platform LLC.
+**
+****************************************************************************/
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "secureaction.h"
+
+#include <QDBusConnection>
+#include <QDBusMessage>
+#include <QDBusObjectPath>
+#include <QDBusPendingCallWatcher>
+#include <QDBusReply>
+#include <QDBusServiceWatcher>
+
+#include <QQmlInfo>
+
+namespace {
+
+const QString deviceLockService = QStringLiteral("org.nemomobile.devicelock");
+const QString authenticatorPath = QStringLiteral("/authenticator");
+const QString authenticatorInterface = QStringLiteral("org.nemomobile.devicelock.Authenticator");
+
+static const QString dbusService = QStringLiteral("org.freedesktop.DBus");
+static const QString dbusPath = QStringLiteral("/org/freedesktop/DBus");
+static const QString dbusInterface = QStringLiteral("org.freedesktop.DBus");
+
+QDBusMessage createAuthenticatorCall(const QString &member, const QVariantList &arguments)
+{
+    QDBusMessage methodCall = QDBusMessage::createMethodCall(
+                deviceLockService,
+                authenticatorPath,
+                authenticatorInterface,
+                member);
+    methodCall.setArguments(arguments);
+    return methodCall;
+}
+
+QString replyPath()
+{
+    static int counter = 0;
+
+    return QStringLiteral("/org/nemomobile/devicelock/client/sailfish_browser/SecureAction%1").arg(++counter);
+}
+
+enum class Method {
+    NoAuthentication    = 0x000,
+    SecurityCode        = 0x0001,
+    Fingerprint         = 0x0002,
+    Confirmation        = 0x1000,
+    AllAvailable = SecurityCode | Fingerprint | Confirmation
+};
+
+}
+
+class SecureAction::ServiceWatcher : public QDBusServiceWatcher
+{
+    Q_OBJECT
+public:
+    ServiceWatcher()
+        : QDBusServiceWatcher(deviceLockService, QDBusConnection::systemBus())
+    {
+        connect(this, &QDBusServiceWatcher::serviceOwnerChanged,
+                this, [this](const QString &, const QString &oldOwner, const QString &newOwner) {
+            if (newOwner.isEmpty() && m_available) {
+                m_available = false;
+                emit availableChanged(false);
+            } else if (oldOwner.isEmpty() && !m_available) {
```

Comment:
Maybe this can never happen in practice, but it looks like if oldOwner and newOwner are both empty and m_available is false, we will enter this branch and set m_available to true here, even though newOwner is still empty.

Reply:
It can't happen, if the old and new owners are the same the owner hasn't changed.